### PR TITLE
Update contains_cftime_datetimes to avoid loading entire variable array

### LINKED ISSUE #7484: Opening datasets with large object dtype arrays is very slow
### What is your issue?

Opening a dataset with a very large array with object dtype is much slower than it should be. I initially noticed this when working with a dataset spanned by around 24000 netcdf files. I have been using kerchunk references to load them with a consolidated metadata key so I was expecting it to be fairly quick to open, but it actually took several minutes. I realized that all this time was spent on one variable consisting of strings, so when dropping it the whole dataset opens up in seconds. Sharing this would be a bit difficult so instead I will illustrate this with a simple easy to reproduce example with the latest released versions of xarray and zarr installed:

```python
str_array = np.arange(100000000).astype(str)
ds = xr.DataArray(dims=('x',), data=str_array).to_dataset(name='str_array')
ds['str_array'] = ds.str_array.astype('O') # Needs to actually be object dtype to show the problem
ds.to_zarr('str_array.zarr')

%time xr.open_zarr('str_array.zarr/')
CPU times: user 8.24 s, sys: 5.23 s, total: 13.5 s
Wall time: 12.9 s
```

I did some digging and found that pretty much all the time was spent on the check being done by `contains_cftime_datetimes` in https://github.com/pydata/xarray/blob/d385e2063a6b5919e1fe9dd3e27a24bc7117137e/xarray/core/common.py#L1793

This operation is not lazy and ends up requiring every single chunk for this variable to be opened, all for the sake of checking the very first element in the entire array. A quick fix I tried is updating `contains_cftime_datetimes` to do the following:
```python
def contains_cftime_datetimes(var) -> bool:
    """Check if an xarray.Variable contains cftime.datetime objects"""
    if var.dtype == np.dtype("O") and var.size > 0:
        ndims = len(var.shape)
        first_idx = np.zeros(ndims, dtype='int32')
        array = var[*first_idx].data
        return _contains_cftime_datetimes(array)
    else:
        return False
```

This drastically reduced the time to open the dataset as expected:
```python
%time xr.open_zarr('str_array.zarr/')
CPU times: user 384 ms, sys: 502 ms, total: 887 ms
Wall time: 985 ms
```

I would like to make a PR with this change but I realize that this change could effect every backend, and although I have been using xarray for many years this would be my first contribution and so I would like to briefly discuss it in case there are better ways to address the issue. Thanks!

### PR DESCRIPTION
**Pull request by agoodm**
- [x] Closes #7484
- [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`

This PR greatly improves the performance for opening datasets with large arrays of object type (typically string arrays) since `contains_cftime_datetimes` was triggering the entire array to be read from the file just to check the very first element in the entire array.

@Illviljan continuing our discussion from the issue thread, I did try to pass in `var._data` to `_contains_cftime_datetimes`, but I had a lot of trouble finding a way to generalize how to index the first array element. The best I could do was `var._data.array.get_array()`, but I don't think `get_array` is implemented for every backend. So for now I am leaving my original proposed solution.

### CODE REVIEW

**Comment by mathause** [comment]
Thanks for the PR. However, does that actually make a difference? To me it looks like `_contains_cftime_datetimes` also only considers one element of the array.

https://github.com/pydata/xarray/blob/b4515582ffc8b7f63632bfccd109d19889d00384/xarray/core/common.py#L1779-L1780

**Comment by agoodm** [comment]
> Thanks for the PR. However, does that actually make a difference? To me it looks like `_contains_cftime_datetimes` also only considers one element of the array.
> 
> https://github.com/pydata/xarray/blob/b4515582ffc8b7f63632bfccd109d19889d00384/xarray/core/common.py#L1779-L1780

This isn't actually the line of code that's causing the performance bottleneck, it's the access to `var.data`  in the function call that is actually problematic as I explained in the issue thread. You can verify this yourself running this simple example before and after applying the changes in this PR:

```python
import numpy as np
import xarray as xr

str_array = np.arange(100000000).astype(str)
ds = xr.DataArray(dims=('x',), data=str_array).to_dataset(name='str_array')
ds['str_array'] = ds.str_array.astype('O') # Needs to actually be object dtype to show the problem
ds.to_zarr('str_array.zarr')

%time xr.open_zarr('str_array.zarr')
```

**Comment by Illviljan** [comment]
@agoodm, what you think of this version? Using xr.Variable directly seems a little easier to work with than trying to guess which type of array (cupy, dask, pint, backendarray, etc) is in the variable.

**Comment by agoodm** [comment]
@Illviljan I gave your update a quick test, it seems to work well enough and still maintains the performance improvement. It looks fine to me though I guess it looks like you still need to fix this failing mypy stuff now?

File: xarray/core/common.py [hunk 1]
```diff
@@ -1771,31 +1772,37 @@ def is_np_timedelta_like(dtype: DTypeLike) -> bool:
...
+def _variable_contains_cftime_datetimes(var: T_Variable) -> bool:
     """Check if an xarray.Variable contains cftime.datetime objects"""
-    if var.dtype == np.dtype("O") and var.size > 0:
-        return _contains_cftime_datetimes(var.data)
-    else:
+    if cftime is None:
         return False
 
+    if var.dtype == np.dtype("O") and var.size > 0:
+        first_idx = tuple(0 for _ in range(var.ndim))
```
Comment (mathause) [inline xarray/core/common.py:1793]:
```suggestion
        first_idx = (0,) * var.ndim
```
to shave off some more ns ;-)

**Review by mathause — COMMENTED** [review]
I thought `da.data` passes the array along as is - but you can learn something everyday. Thanks @Illviljan for taking over and sorry @agoodm for not properly reading the issue...

File: xarray/core/common.py [hunk 2]
```diff
@@ -1771,31 +1772,37 @@ def is_np_timedelta_like(dtype: DTypeLike) -> bool:
...
+        first_idx = (0,) * var.ndim
+        sample = var[first_idx]
+        return isinstance(sample.to_numpy().item(), cftime.datetime)
+
+    return False
+
+
+def contains_cftime_datetimes(var: T_Variable) -> bool:
+    """Check if an xarray.Variable contains cftime.datetime objects"""
+    return _variable_contains_cftime_datetimes(var)
```
Comment (dcherian) [inline xarray/core/common.py:1802]: Can we remove this?

File: xarray/core/common.py [hunk 3]
```diff
@@ -1771,31 +1772,37 @@ def is_np_timedelta_like(dtype: DTypeLike) -> bool:
...
-    if var.dtype == np.dtype("O") and var.size > 0:
-        return _contains_cftime_datetimes(var.data)
-    else:
+    if cftime is None:
         return False
 
+    if var.dtype == np.dtype("O") and var.size > 0:
+        first_idx = (0,) * var.ndim
+        sample = var[first_idx]
+        return isinstance(sample.to_numpy().item(), cftime.datetime)
```
[the remark this answers is not in the document]
Comment (agoodm) [reply xarray/core/common.py:1794]: Thanks for taking a look at this. I am a little confused for what you are suggesting here. Are you looking for a simple test in `test_variable.py` that applies the same logic in this block to extract the very first element via `Variable.__getitem__` here and check that it returns one value, a more general `contains_cftime_datetimes` test, or both?

File: doc/whats-new.rst [hunk 4]
```diff
@@ -99,6 +99,8 @@ Bug fixes
 - :py:func:`xarray.Dataset.to_zarr` now drops variable encodings that have been added by xarray during reading
   a dataset. (:issue:`7129`, :pull:`7500`).
   By `Hauke Schulz <https://github.com/observingClouds>`_.
+- Improved performance in ``open_dataset`` for datasets with large object arrays (:issue:`7484`, :pull:`7494`).
```
Comment (headtr1ck) [inline doc/whats-new.rst:102]: Move to new section

**Review by headtr1ck — COMMENTED** [review]
LGTM.

Some simple test of the new functionality would be nice.

If you are really motivated, you can think about adding an asv benchmark here: https://github.com/pydata/xarray/blob/main/asv_bench/benchmarks/dataset_io.py

File: xarray/core/common.py [hunk 5]
```diff
@@ -1770,31 +1771,32 @@ def is_np_timedelta_like(dtype: DTypeLike) -> bool:
...
 
 
-def _contains_cftime_datetimes(array) -> bool:
+def _contains_cftime_datetimes(array: Any) -> bool:
     """Check if an array contains cftime.datetime objects"""
-    if cftime is None:
-        return False
+    from xarray.core.variable import Variable
+
+    if isinstance(array, Variable):
```
Comment (dcherian) [inline xarray/core/common.py:1778]:
```suggestion
    if not isinstance(array, Variable):
```

File: xarray/core/common.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (dcherian) [reply xarray/core/common.py]: Sotry that was a bit complicated and intended for IIlviljan.

I pushed a commit with a test. I also changed the code to account for those lazily  indexed backend arrays explicitly.

**Comment by dcherian** [comment]
Seems like we're passing  a DataArray instead of a Variable somewhere.

**Comment by mathause** [comment]
I think that's in the tests themselves

https://github.com/pydata/xarray/blob/6531b57f8c5cb7f3c564ff895c2e4b6573bb5521/xarray/tests/test_coding_times.py#L778

File: xarray/core/common.py [hunk 6]
```diff
@@ -1770,31 +1772,29 @@ def is_np_timedelta_like(dtype: DTypeLike) -> bool:
...
+    return False
 
-def contains_cftime_datetimes(var) -> bool:
+
+def contains_cftime_datetimes(var: T_Variable) -> bool:
     """Check if an xarray.Variable contains cftime.datetime objects"""
-    if var.dtype == np.dtype("O") and var.size > 0:
-        return _contains_cftime_datetimes(var.data)
-    else:
+    if cftime is None:
```
Comment (Illviljan) [inline xarray/core/common.py:1792]: Can remove this check. It is done in `_contains_cftime_datetimes` as well.

File: xarray/tests/test_coding.py [hunk 7]
```diff
@@ -65,13 +65,13 @@ def test_CFMaskCoder_missing_value() -> None:
     expected.attrs["missing_value"] = -9999
 
     decoded = xr.decode_cf(expected.to_dataset())
-    encoded, _ = xr.conventions.cf_encoder(decoded, decoded.attrs)
+    encoded, _ = xr.conventions.cf_encoder(decoded.variables, decoded.attrs)
```
Comment (dcherian) [inline xarray/tests/test_coding.py:68]: @Illviljan This looks like a typing opportunity for a future PR.

Comment (Illviljan) [reply xarray/tests/test_coding.py:68]: Sooo many to choose from! :D

**Review by Illviljan — APPROVED** [review]

**Comment by dcherian** [comment]
Thanks @agoodm this work prompted a bunch of internal cleanup!

**Comment by agoodm** [comment]
Thanks @Illviljan and @dcherian for helping to see this through.

**Pull request merged**